backend/session/session.c

```diff
@@ -533,3 +533,59 @@ ssize_t wlr_session_find_gpus(struct wlr_session *session,
 
 	return i;
 }
+
+struct wait_session_active_handler {
+	bool active;
+	struct wl_listener listener;
+};
+
+static void	wait_session_active(struct wl_listener *listener, void *data) {
+	struct wait_session_active_handler *handler =
+		wl_container_of(listener, handler, listener);
+	handler->active = true;
+}
+
+struct wlr_session *wlr_session_create(struct wl_display *disp) {
+	struct wlr_session *session = _wlr_session_create(disp);
+
+	if (!session) {
+		wlr_log(WLR_ERROR, "Failed to start a session");
+		return NULL;
+	}
+
+	if (!session->active) {
+		wlr_log(WLR_INFO, "Waiting for a session to become active");
+
+		struct wait_session_active_handler handler = {0};
+		handler.listener.notify = wait_session_active;
+		wl_signal_add(&session->events.active, &handler.listener);
+
+		uint64_t started_at = get_current_time_ms();
+		uint64_t timeout = WAIT_GPU_TIMEOUT;
+		struct wl_event_loop *event_loop =
+			wl_display_get_event_loop(session->display);
+		while (!handler.active) {
```

Comment:
Instead of adding a listener, we can just have `while (session->active)` here

Reply:
@emersion, actualy i'd better not loop here, instead I would pass longer timeout, and exit right after it, like this:

```
struct wlr_session *wlr_session_create(struct wl_display *disp) {
	struct wlr_session *session = _wlr_session_create(disp);

	if (!session) {
		wlr_log(WLR_ERROR, "Failed to start a session");
		return NULL;
	}

	if (!session->active) {
		wlr_log(WLR_INFO, "Waiting for a session to become active");

		struct wait_session_active_handler handler = {0};
		handler.listener.notify = wait_session_active;
		wl_signal_add(&session->events.active, &handler.listener);

		struct wl_event_loop *event_loop =
			wl_display_get_event_loop(session->display);

		int ret = wl_event_loop_dispatch(event_loop, (int)WAIT_SESSION_TIMEOUT);
		if (ret < 0) {
			wlr_log_errno(WLR_ERROR, "Failed to wait for session active: "
						  "wl_event_loop_dispatch failed");
			return NULL;
		}

		wl_list_remove(&handler.listener.link);

		if (!handler.active) {
			wlr_log(WLR_ERROR, "Wait session active timeout");
			return NULL;
		}
	}

	return session;
}
```
Where `WAIT_SESSION_TIMEOUT` is 30 seconds or so.. Session either will come or not, loop is unecessary i think, and more clean, now this function can be moved next to original `wlr_session_create`.

Reply:
No, a loop is needed here. `wl_event_loop_dispatch` can return at an arbitrary time, if an event comes in it's not necessarily a libseat one.